src/icp_canister_backend/src/lib.rs

```diff
@@ -57,43 +58,73 @@ thread_local! {
             }
         ).expect("Failed to initialize PoolState")
     );
+
+    static EPISODES: RefCell<StableBTreeMap<u64, Episode, Memory>> = RefCell::new(
+        StableBTreeMap::init(
+            MEMORY_MANAGER.with(|m| m.borrow().get(MemoryId::new(5))),
+        )
+    );
+}
+
+fn get_current_episode() -> u64 {
+    ic_cdk::api::time() / 1_000_000_000 / *EPISODE_DURATION
+}
+
+fn is_episode_active(episode_id: u64) -> bool {
+    let current_episode = get_current_episode();
+    episode_id >= current_episode && episode_id < current_episode + *MAX_ACTIVE_EPISODES
 }
 
 #[ic_cdk::query]
-pub fn get_deposit_subaccount(user: Principal, timelock: u64) -> [u8; 32] {
+pub fn get_deposit_subaccount(user: Principal, episode: u64) -> [u8; 32] {
     let mut hasher = Sha256::new();
     hasher.update(user.as_slice());
-    hasher.update(timelock.to_be_bytes());
+    hasher.update(episode.to_be_bytes());
     hasher.finalize().into()
 }
 
+#[ic_cdk::query]
+pub fn get_current_episode_id() -> u64 {
+    get_current_episode()
+}
+
+#[ic_cdk::query]
+pub fn get_episode(episode_id: u64) -> Option<Episode> {
+    EPISODES.with(|episodes| episodes.borrow().get(&episode_id))
+}
+
 #[ic_cdk::query]
 pub fn get_user_deposits(user: Principal) -> Vec<types::UserDepositInfo> {
     let deposit_ids = USER_DEPOSITS.with(|user_deposits| match user_deposits.borrow().get(&user) {
         Some(deposits) => deposits.0.clone(),
         None => return Vec::new(),
     });
 
-    let pool_state = POOL_STATE.with(|state| state.borrow().get().clone());
-
     DEPOSITS.with(|deposits| {
         let deposits_ref = deposits.borrow();
         deposit_ids
             .iter()
             .filter_map(|&deposit_id| {
                 deposits_ref.get(&deposit_id).map(|deposit| {
-                    let amount = if pool_state.total_shares == Nat::from(0u128) {
+                    let episode_data = EPISODES.with(|episodes| {
```

Comment:
It is not good to have such an operation in a loop - calling EPISODES.with(|episodes|.
Call this only once.